### DISCUSSION TITLE
How to integrate atomic swap functionality into wallet?

### DISCUSSION THREAD

**Discussion by nmalchy** [comit-network/xmr-btc-swap#1027] [Q&A] [answered]
Hello,

I am wondering if it if possible to implement your atomic swap MVP into an existing or new mobile wallet, so let's say the wallet supports Bitcoin and Monero, how difficult do you think it would be to build in the ability to retrieve a list of swap providers, select x provider, send btc to the swap provider and the monero address for the provider to send the XMR to and then maybe return a successful tx hash? Sorry I'm a bit unfamiliar with the technical minutia of how this atomic swap thing works, maybe I'm asking for a lot, I would be willing to work with the team to help achieve this, thanks for your amazing work!!

**Comment by delta1** [comment, marked as answer]
The short answer is, it would be a lot of work. Your wallet needs to be able to talk `libp2p`, which there is probably an implementation (or binding) for, but then you'd have to reimplement the logic of the [swap CLI](https://github.com/comit-network/xmr-btc-swap/blob/master/swap/src/bin/swap.rs) in your wallet... 

It would be easier if the swap event loop were in some library you could just import, but that's quite a task and this project is currently run by unfunded volunteer maintainers.
